ddht/tools/lru_sql_dict.py

```diff
@@ -0,0 +1,269 @@
+
+    def __delitem__(self, key: TKey) -> None:
+        serialized_key = self.key_encoder(key)
+        result = self._fetch_single_query(
+            "SELECT key,value,pref,nref FROM cache WHERE key=?;", (serialized_key,),
+        )
+
+        if not result:
+            raise KeyError(str(key))
+
+        node = Node(*result)
+
+        # delete key from cache
+        self._execute(
+            "DELETE FROM cache WHERE key=?;", (node.key,),
+        )
+
+        # update any nrefs/prefs in cache
+        nref_node = self._fetch_single_query(
+            "SELECT key,value,pref,nref FROM cache WHERE nref=?;", (node.key,),
+        )
+        pref_node = self._fetch_single_query(
+            "SELECT key,value,pref,nref FROM cache WHERE pref=?;", (node.key,),
+        )
+
+        if nref_node and pref_node:
+            nref = Node(*nref_node)
+            pref = Node(*pref_node)
+            self._execute(
+                "UPDATE cache SET nref=? WHERE key=?;", (pref.key, nref.key),
+            )
+            self._execute(
+                "UPDATE cache SET pref=? WHERE key=?;", (nref.key, pref.key),
+            )
+        elif nref_node:
+            nref = Node(*nref_node)
+            self._execute(
+                "UPDATE cache SET nref=? WHERE key=?;", (None, nref.key),
+            )
+        elif pref_node:
+            pref = Node(*pref_node)
+            self._execute(
+                "UPDATE cache SET pref=? WHERE key=?;", (None, pref.key),
+            )
+
+    def _insert_item(self, key: TKey, value: TValue) -> None:
+        # insert new item into map and cache
+        serialized_key = self.key_encoder(key)
+        serialized_value = self.value_encoder(value)
+        if self.is_empty:
+            new_pref = None
+            new_nref = None
+
+            self._execute(
+                "INSERT INTO cache VALUES (?, ?, ?, ?);",
+                (serialized_key, serialized_value, new_pref, new_nref),
+            )
+
+        else:
+            # evict least recently used key/value
+            if self.is_full:
+                self.__delitem__(self.key_decoder(self.tail.key))
+
+            # get old head
+            old_head_key, _, old_head_pref, _ = self.head
+
+            # add new head to cache
+            self._execute(
+                "INSERT INTO cache VALUES (?, ?, ?, ?);",
+                (serialized_key, serialized_value, None, old_head_key),
+            )
+
+            # update old head in cache
+            self._execute(
+                "UPDATE cache SET pref=? WHERE key=?;", (serialized_key, old_head_key),
+            )
+
+    def _update_item(self, key: TKey, value: TValue) -> None:
+        serialized_key = self.key_encoder(key)
+        serialized_value = self.value_encoder(value)
+        self._execute(
+            "UPDATE cache SET value=? WHERE key=?;", (serialized_value, serialized_key)
+        )
+
+    def _fetch_single_query(self, query: str, args: Tuple[Any, ...] = None) -> Any:
```

Comment:
Since tuples are immutable, you can *in theory* change this to just be:

```
     def _fetch_single_query(self, query: str, args: Tuple[Any, ...] = ()) -> Any: 
```

This should let you remove the `if/else` below since it appears that `conn.execute` is happy to have an empty tuple passed in for `args` when the query has no args.